### PR TITLE
Add processing for ICLR-GREAT Paper and Fix bugs

### PR DESCRIPTION
**Pull request by hsellik**
This pull request adds DataGenerator for [ICLR2020-Great](https://github.com/VHellendoorn/ICLR20-Great) paper and also fixes some bugs.

2 Additional tests were added. One for unparsable code (expecting an illegalArgumentException) and one for Enums.

The enum test has a @TODO tag since j2graph currently fails to handle them.

### CODE REVIEW

File: src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Tests that exercise the classes with Enums.
+ * This was a bug (issue #3)
+ */
+public class EnumTest extends TestBase {
+
+//    private ClassGraph graph = run("representation/Enum.java");
+
+    //    @TODO - this test fails at the moment. Graph should not be null with enums.
```
Comment (mauricioaniche) [inline src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java:15]: Is this failing? If so, maybe adding a `@Ignore` or something?

File: src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * Visitor to help generate data for ICLR20-Great data format: https://github.com/VHellendoorn/ICLR20-Great
+ * <p>
+ * The Visitor also mutates binary expressions (<,<=,>,>=)
+ */
+public class JsonVisitor implements Walker {
```
Comment (mauricioaniche) [inline src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java]: Maybe we should call this class `OffSideVisitor`, so that we fully couple it to your work? Otherwise, this might confuse others.

File: src/main/java/com/github/sergdelft/j2graph/DataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
+package com.github.sergdelft.j2graph;
```
Comment (mauricioaniche) [inline src/main/java/com/github/sergdelft/j2graph/DataGenerator.java]: can we put this in a `iclr20great` package?

File: src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java (hunk 1 above)
Comment (mauricioaniche) [reply src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java:15]: And open an issue to it?

File: src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.github.sergdelft.j2graph.TestBase;
+import com.github.sergdelft.j2graph.ast.JDT;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Tests that exercise the classes with imports.
+ * This was a bug (issue #3)
```
Comment (mauricioaniche) [inline src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java]: is this really related to issue #3 ?

**Review by mauricioaniche — CHANGES_REQUESTED** [review]
Thanks for the PR! Some feedback!

File: src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java (hunk 4 above)
Comment (hsellik) [reply src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java]: Oh, that's a copy-paste error. Will fix.

File: src/main/java/com/github/sergdelft/j2graph/DataGenerator.java (hunk 3 above)
Comment (hsellik) [reply src/main/java/com/github/sergdelft/j2graph/DataGenerator.java]: Will do

File: src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java (hunk 2 above)
Comment (hsellik) [reply src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java]: Makes sense, I was thinking the same thing about the JsonVisitor being ambiguous actually.

File: src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java (hunk 1 above)
Comment (hsellik) [reply src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java:15]: Yes, it is failing, I will open an issue. Edit: already opened an issue actually: https://github.com/SERG-Delft/j2graph/issues/7

File: src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java (hunk 2 above)
Comment (hsellik) [reply src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java]: @mauricioaniche , perhaps ICLR20GreatVisitor is better?

**Comment by hsellik** [comment]
@mauricioaniche , made the changes.

File: src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java (hunk 4 above)
Comment (hsellik) [reply src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java]: Done

File: src/main/java/com/github/sergdelft/j2graph/DataGenerator.java (hunk 3 above)
Comment (hsellik) [reply src/main/java/com/github/sergdelft/j2graph/DataGenerator.java]: Done

File: src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java (hunk 2 above)
Comment (hsellik) [reply src/main/java/com/github/sergdelft/j2graph/walker/json/JsonVisitor.java]: Done, named to ICLR20GreatVisitor and also named the DataGenerator to ICLR20GreatDataGenerator.

File: src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java (hunk 1 above)
Comment (hsellik) [reply src/test/java/com/github/sergdelft/j2graph/representation/EnumTest.java:15]: Done

File: src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java (hunk 4 above)
Comment (mauricioaniche) [reply src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java]: I still see 3 here?

File: src/main/java/com/github/sergdelft/j2graph/ast/JDTVisitor.java [hunk 5]
```diff
@@ -106,7 +106,7 @@ public boolean visit(SimpleName node) {
         // however, it might be that we are not inside a method
         // e.g., field declaration
         // so, we only collect it if we are inside a method
-        if (inAMethod()) {
+        if (inAMethod() && !nonTerminals.get(currentMethod()).isEmpty()) {
```
Comment (mauricioaniche) [inline src/main/java/com/github/sergdelft/j2graph/ast/JDTVisitor.java:109]: clarification question: why did you have to add this condition everywhere?

I thought the `inAMethod()` was handling it already.

File: src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java (hunk 4 above)
Comment (hsellik) [reply src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java]: Oops, I fixed it in the Enum test. 🤦 Pushed for the unparsable code test as well now.

File: src/main/java/com/github/sergdelft/j2graph/ast/JDTVisitor.java (hunk 5 above)
Comment (hsellik) [reply src/main/java/com/github/sergdelft/j2graph/ast/JDTVisitor.java:109]: `inAMethod()` checks for `methodBuilders` and `nonTerminals` variables. However, if I don't add the check, the call in the next line for `currentNonTerminal()` will call `nonTerminals.get(currentMethod()).peek()` will result in a peek error. So the `nonTerminals` variable is not empty, but the `nonTerminals.get(currentMethod()).peek()` will still fail.

Also, the reason I mentioned earlier that my fixes are a little bit hacked. I added a lot of checks in order to avoid those errors. However, I don't know the actual reason why the peek in `currentNonTerminal()` method will fail. I can probably give you example methods if you would like to debug.

File: src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java (hunk 4 above)
Comment (hsellik) [reply src/test/java/com/github/sergdelft/j2graph/representation/UnparsableCodeTest.java]: Oops, I fixed the Enum test that had the same error. Added the commit for UnparsableCode. I guess the IllegalArgumentException is the expected output for unparsable code?

File: src/main/java/com/github/sergdelft/j2graph/ast/JDTVisitor.java (hunk 5 above)
Comment (mauricioaniche) [reply src/main/java/com/github/sergdelft/j2graph/ast/JDTVisitor.java:109]: If you have one example to send me, that'd be nice. Then, later I can think of the cleanest solution!
